### PR TITLE
feat/LLM::Openai#chat : `stream:` option

### PR DESCRIPTION
**Pull request by mael-ha**
## Context
Hey  ✌️
Using this gem for a Rails + React app as a Chatbot, I wanted to be able to stream ChatGPT's answer directly to the front end. [This feature](https://github.com/alexrudall/ruby-openai#streaming-chatgpt) is available in the `openai-ruby` gem, so made a small modification to be able to use it in `langchainrb`

## Solution
- added the `stream:` param to `#chat`, (`boolean`, default: `false`), when set to `true` we should then provide a block with what has to be performed in the stream proc. In my case, for instance broadcasting with ActionCable or use turbo stream or whatever. 

```
content = ''
@llm_client.chat(prompt:, stream: true) do |new_content|
   content += new_content if new_content
   ActionCable.server.broadcast('AnswersChannel', { content:, streamStopped: new_content ? false : true })
end
```

### CODE REVIEW

File: lib/langchain/llm/openai.rb [hunk 1]
```diff
@@ -64,15 +64,29 @@ def complete(prompt:, **params)
...
+    #
+    def chat(prompt:, stream: false, **params)
       parameters = compose_parameters DEFAULTS[:chat_completion_model_name], params
 
       Langchain::Utils::TokenLengthValidator.validate!(prompt, parameters[:model])
 
-      parameters[:messages] = [{role: "user", content: prompt}]
+      parameters[:messages] = [{ role: 'user', content: prompt }]
+
+      if stream
```
[the remark this answers is not in the document]
Comment (andreibondarev) [reply lib/langchain/llm/openai.rb:82]: Cool! I was able to validate it with:
```ruby
irb(main):015:1* openai.chat(prompt: "Hello! How are you?", stream: true) do |res|
irb(main):016:1*   print res
irb(main):017:0> end
```

Maybe we should validate that a block is passed in when `stream == true`?

File: lib/langchain/llm/openai.rb [hunk 2]
```diff
@@ -66,20 +66,29 @@ def complete(prompt:, **params)
...
     #
-    def chat(prompt: "", messages: [], context: "", examples: [], **options)
+    def chat(prompt: "", messages: [], context: "", examples: [], stream: false, **options)
       raise ArgumentError.new(":prompt or :messages argument is expected") if prompt.empty? && messages.empty?
 
       parameters = compose_parameters DEFAULTS[:chat_completion_model_name], options
       parameters[:messages] = compose_chat_messages(prompt: prompt, messages: messages, context: context, examples: examples)
       parameters[:max_tokens] = validate_max_tokens(parameters[:messages], parameters[:model])
 
+      if (streaming = stream && block_given?)
```
Comment (alchaplinsky) [inline lib/langchain/llm/openai.rb:79]: I think we can avoid having `stream` option and just look at `block_given?` So we'll call this method either like:

```ruby
result = llm.chat(prompt: "hello")
```

or 
```ruby
llm.chat(prompt: "hello") do |chunk|
   # do stuff
end
```

File: lib/langchain/conversation.rb [hunk 3]
```diff
@@ -19,12 +24,13 @@ class Conversation
     # @param llm [Object] The LLM to use for the conversation
     # @param options [Hash] Options to pass to the LLM, like temperature, top_k, etc.
     # @return [Langchain::Conversation] The Langchain::Conversation instance
-    def initialize(llm:, **options)
+    def initialize(llm:, **options, &block)
```
[the remark this answers is not in the document]
Comment (andreibondarev) [reply lib/langchain/conversation.rb:27]: You could make an argument that `**options` should be passed along to `message()` as well. 

Let's keep it here for now, but something to potentially reconsider in the future.

**Comment by andreibondarev** [comment]
@mael-ha Thank you for starting this PR, I took it across the finish line! 🏃🏻‍♂️🏁 🎉

**Comment by mael-ha** [comment]
> @mael-ha Thank you for starting this PR, I took it across the finish line! 🏃🏻‍♂️🏁 🎉

🙏 Thanks, you rock! Was kinda busy the last days :D

**Pull request merged**